aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h

```diff
@@ -0,0 +1,309 @@
+// for all i = 1, ..., compressed_dim
+// are sorted and distinct along the last dimension values.
+template <CDimName cdim_name, typename index_t>
+INVARIANT_CHECK_FUNC_API
+_check_idx_sorted_distinct_vals_slices_with_cidx(
+    const index_t* RESTRICT ptr_idx_batch,
+    const index_t cidx,
+    const index_t cidx_next) {
+  // Note that ptr_idx_batch = &idx[batch_idx] and is contiguous.
+  const auto* RESTRICT slice_begin = ptr_idx_batch + cidx;
+  const auto* RESTRICT slice_end = ptr_idx_batch + cidx_next;
+  for (auto* RESTRICT curr = slice_begin + 1; curr < slice_end; ++curr) {
+    const auto invariant = *(curr - 1) < *curr;
+    if (cdim_name == CDimName::CRow) {
+      _assert(invariant, "`col_indices[..., crow_indices[..., i - 1]:crow_indices[..., i]] "
+                         "for all i = 1, ..., nrows "
+                         "are sorted and distinct along the last dimension values` "
+                         "is not satisfied.");
+    }
+    else {
+      _assert(invariant, "`row_indices[..., ccol_indices[..., i - 1]:ccol_indices[..., i]] "
+                         "for all i = 1, ..., ncols "
+                         "are sorted and distinct along the last dimension values` "
+                         "is not satisfied.");
+    }
+  }
+}
+
+static inline int64_t numel(IntArrayRef sizes) {
+  int64_t res = 1;
+  for (const auto& s : sizes) {
+    res *= s;
+  }
+  return res;
+}
+
+template <typename func_t, typename vec_func_t>
+struct EmptyVecKernel {
+  static void launch(TensorIteratorBase& iter, const func_t& f, const vec_func_t& vec_f) {
+  }
+};
+
+template <typename scalar_t>
+using DummyVec = scalar_t;
+
+template <
+  template <typename func_t> class kernel_t,
+  template <typename func_t, typename vec_func_t> class vec_kernel_t>
+struct KernelLauncher {
+  template <typename func_t, typename vec_func_t>
+  static void launch(TensorIteratorBase& iter, const func_t& f, const vec_func_t& vec_f) {
+    vec_kernel_t<func_t, vec_func_t>::launch(iter, f, vec_f);
+  }
+
+  template <typename func_t>
+  static void launch(TensorIteratorBase& iter, const func_t& f) {
+    kernel_t<func_t>::launch(iter, f);
+  }
+};
+
+template <
+  CDimName cdim_name,
+  template <typename func_t> class kernel_t,
+  template <typename func_t, typename vec_func_t> class vec_kernel_t = EmptyVecKernel,
+  template <typename scalar_t> class Vec = DummyVec>
+void _validate_compressed_sparse_indices_kernel(
+    const Tensor& cidx,
+    const Tensor& idx,
+    const int64_t cdim,
+    const int64_t dim,
+    const int64_t nnz) {
+  TORCH_CHECK(cidx.size(-1) == cdim + 1, "c{row|col}_indices have wrong shape: ",
+      "c{row|col}.shape[-1] = ", cidx.size(-1), " is not equal to ",
+      "cdim + 1 = ", cdim + 1);
+  TORCH_CHECK(idx.size(-1) == nnz, "{row|col}_indices have wrong shape: ",
+      "{row|col}.shape[-1] = ", idx.size(-1), " is not equal to ",
+      "nnz = ", nnz);
```

Comment:
Ah, totally missed, good catch!

Reply:
BTW, did it compile? `if constexpr` is a feature of C++17, or have we switched already?

Reply:
I didn't try to compile it. Actually, a compiler might be able to optimize the if-block away (because the condition is template parameter dependent only) so that `constexpr` might not be needed.

Reply:
I am removing these as they are checked in the enclosing function and are tested already.

Reply:
Yeah, it is not needed and it should optimize it out, the whole point of having template arguments ;)